WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,626 @@
+                }
+            } else if nPrefixes == 1 {
+                streamCopy.appendIfNew(key: String(s[0]), value: stream)
+            }
+        }
+
+        #if DEBUG
+        do {
+            let jsonString = try streamCopy.toPrettyPrintJSONString()
+            DDLogDebug("[EPC] Map of streams to CC:\n\(jsonString)")
+        } catch let error {
+            DDLogError("[EPC] \(error.localizedDescription)")
+        }
+        #endif
+
+        queue.async {
+            if self.inputBuffer.count > 0 {
+                var cachedEvent: Event
+                while !self.inputBuffer.isEmpty {
+                    cachedEvent = self.inputBuffer.remove(at: 0)
+                    self.log(stream: cachedEvent.stream,
+                             schema: cachedEvent.schema,
+                             data: cachedEvent.data,
+                             domain: cachedEvent.domain)
+                }
+            }
+        }
+    }
+
+    /**
+     * Yields a deterministic (not stochastic) determination of whether the provided `id` is
+     * in-sample or out-of-sample according to the `acceptance` rate
+     * - Parameter id: either session ID generated with `generateID` or the app install ID
+     * generated with `UUID().uuidString`
+     * - Parameter acceptance: the desired proportion of many `token`-s being accepted
+     *
+     * The algorithm works in a "widen the net on frozen fish" fashion -- tokens continue evaluating to
+     * true as the acceptance rate increases. For example, a device determined to be in-sample for a
+     * stream "A" having rate 0.1 will be determined to be in-sample for a stream "B" having rate 0.2,
+     * and its events will show up in tables "A" and "B".
+     */
+    private func determine(_ id: String, _ acceptance: Double) -> Bool {
+        guard let token = UInt32(id.prefix(8), radix: 16) else {
+            return false
+        }
+        return (Double(token) / Double(UInt32.max)) < acceptance
+    }
+
+    /**
+     * Compute a boolean function on a random identifier
+     * - Parameter stream: name of the stream
+     * - Returns: `true` if in sample or `false` otherwise
+     *
+     * The determinations are lazy and cached, so each stream's in-sample vs out-of-sample determination
+     * is computed only once, the first time an event is logged to that stream.
+     *
+     * Refer to sampling settings section in
+     * [mw:Wikimedia Product/Analytics Infrastructure/Stream configuration](https://www.mediawiki.org/wiki/Wikimedia_Product/Analytics_Infrastructure/Stream_configuration)
+     * for more information.
+     */
+    private func inSample(stream: String) -> Bool {
+
+        guard let configs = streamConfigurations else {
+            DDLogDebug("[EPC] Invalid state, must have streamConfigurations to check for inSample")
+            return false
+        }
+        if let cachedValue = samplingCache[stream] {
```

Comment:
Actually this piece will need some thread-safety too. For this simple serial queue approach you can try `queue.sync`, that pauses the current thread execution while `queue` goes off and pulls the value and returns it back to the original thread that needs it. A pattern I've noticed in this codebase is to create a bunch of helper methods for getting and setting to these data structures. That buries all the queue.async and queue.sync calls and cleans up the code. It was a bit long to give an example here so I tried it in this [commit](https://github.com/wikimedia/wikipedia-ios/commit/36fc152c3edbd2fff651e8a9f9c556e5a7981ea0).

Also as I'm looking at this I think we might need thread safety when accessing or setting the `streamConfigurations` and `streamCopy` properties as well 😕   so I'll leave that for you. The reason being they are referenced in `log(...)`, which I envision will get called from multiple threads as we pepper them throughout our codebase, so we want to make sure these properties stay intact.

Reply:
Thanks! Hopefully I did it right in the upcoming commit